### PR TITLE
Ledger-API: Validate Dar before sending it to the ledger.

### LINKED ISSUE #7399: Ledger: Validate Packages as part of the package upload
Currently packages are validated by the Engine before Speedy compilation, but not when uploaded to the Ledger.  

The package should be validated before to be stored on the Ledger as part of the package upload process.
 This has two main advantages: 
* this avoids to store bogus data on the ledger that cannot be latter used. 
* this allowed to skip validation in the Engine, then speeding up submission when packages are not available in the Engine. 

# Proposal 

1 - The participant Node validate the packages before sending them to the ledger.  
     This prevents the ledger to validated buggy archive, since this can be very costly in a distributed system.
2 - The ledger revalidate the package before committing, reject if invalid   
3 - Drop the validation in the Engine.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR implements a part of the proposal from #7093.

Here packages are validated in the participant node before to be sent to the ledger. 

CHANGELOG_BEGIN

- [Ledger-API] participant node validate Dar before uploading to the ledger.
  This may increase upload time significantly.

CHANGELOG_END

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [X] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala [hunk 1]
```diff
@@ -15,7 +15,7 @@ import scala.concurrent.{Await, ExecutionContext, Future}
 class AkkaExecutionSequencerPool(
     poolName: String,
     actorCount: Int = AkkaExecutionSequencerPool.defaultActorCount,
-    terminationTimeout: FiniteDuration = 30.seconds,
+    terminationTimeout: FiniteDuration = 180.seconds,
```
Comment (SamirTalwar-DA) [inline ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala:18]: I think you also need to increase the timeout in [ApiPackageManagementService.scala](https://github.com/digital-asset/daml/blob/d48e9d251d22ba0aea3de2cd51755e3be447e8a1/ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala#L56).

**Review by SamirTalwar-DA — COMMENTED** [review]
Much nicer than before!

File: ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala:18]: Do I actually need it here ?

Comment (SamirTalwar-DA) [reply ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala:18]: I don't think so, but I might be wrong. I suggest testing it with a very large DAR.

**Review by SamirTalwar-DA — APPROVED** [review]

File: daml-lf/validation/BUILD.bazel [hunk 2]
```diff
@@ -13,10 +13,7 @@ da_scala_library(
     srcs = glob(["src/main/**/*.scala"]),
     scalacopts = lf_scalacopts,
     tags = ["maven_coordinates=com.daml:daml-lf-validation:__VERSION__"],
-    visibility = [
-        "//compiler/scenario-service:__subpackages__",
-        "//daml-lf:__subpackages__",
-    ],
+    visibility = ["//visibility:public"],
```
Comment (stefanobaghino-da) [inline daml-lf/validation/BUILD.bazel:16]:
```suggestion
    visibility = ["//:__subpackages__"],
```
Unless you are interested in this being globally visible.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala [hunk 3]
```diff
@@ -79,22 +80,29 @@ private[apiserver] final class ApiPackageManagementService private (
...
+    } yield dar
+
   override def uploadDarFile(request: UploadDarFileRequest): Future[UploadDarFileResponse] = {
     val submissionId =
       if (request.submissionId.isEmpty)
         SubmissionId.assertFromString(UUID.randomUUID().toString)
       else
         SubmissionId.assertFromString(request.submissionId)
 
+    val stream = new ZipInputStream(new ByteArrayInputStream(request.darFile.toByteArray))
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:99]: Do any of those streams need to be explicitly closed?

Comment (remyhaemmerle-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala:99]: Good question. It seems that `DarReader#readArchive` close the stream given in argument.

```scala

  /** Reads an archive from a ZipInputStream. The stream will be closed by this function! */
  def readArchive(
      name: String,
      darStream: ZipInputStream,
      entrySizeThreshold: Int = EntrySizeThreshold,
  ): Try[Dar[A]] = ...
```

**Comment by cocreature** [comment]
Fabric hits the timeout pretty regularly both on CI as well as by users (e.g., https://discuss.daml.com/t/daml-on-fabric-deadline-exceeded-during-daml-deploy/1252) so in theory it would be a great testcase to make sure that you bumped all the right places. Unfortunately, Fabric is also not even upgraded to 1.4.0 (because some APIs broke) so unfortunately I don’t how how you could reasonably test this.

File: daml-lf/validation/BUILD.bazel (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/BUILD.bazel:16]: As part of #7399, each ledger implementation will have to validate packages before committing them.
So they will need access to this packages.

File: ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/AkkaExecutionSequencerPool.scala:18]: reverted.

**Pull request merged**